src/app/functions/server/getRelevantGuidedSteps.ts

```diff
@@ -0,0 +1,260 @@
+
+import { Session } from "next-auth";
+import { LatestOnerepScanData } from "../../../db/tables/onerep_scans";
+import { SubscriberBreach } from "../../../utils/subscriberBreaches";
+import { BreachDataTypes, HighRiskDataTypes } from "../universal/breach";
+
+export type InputData = {
+  user: Session["user"];
+  countryCode: string;
+  latestScanData: LatestOnerepScanData | null;
+  subscriberBreaches: SubscriberBreach[];
+};
+
+// Note: the order is important; it determines in which order the user will be
+//       guided through the pages.
+export const stepLinks = [
+  {
+    href: "/redesign/user/dashboard/fix/data-broker-profiles/start-free-scan",
+    id: "StartScan",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/data-broker-profiles/view-data-brokers",
+    id: "ScanResult",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/social-security-number",
+    id: "HighRiskSsn",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/credit-card-number",
+    id: "HighRiskCreditCard",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/bank-account",
+    id: "HighRiskBankAccount",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/pin-number",
+    id: "HighRiskPin",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/leaked-passwords/password",
+    id: "LeakedPasswordsPassword",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/leaked-passwords/security-question",
+    id: "LeakedPasswordsSecurityQuestion",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/phone",
+    id: "SecurityTipsPhone",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/email",
+    id: "SecurityTipsEmail",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/ip",
+    id: "SecurityTipsIp",
+  },
+  {
+    href: "/redesign/user/dashboard",
+    id: "Done",
+  },
+] as const satisfies ReadonlyArray<{ href: string; id: string }>;
+
+export type StepLink = (typeof stepLinks)[number];
+export type StepLinkWithStatus = (typeof stepLinks)[number] & {
+  eligible: boolean;
+  completed: boolean;
+};
+
+export type OutputData = {
+  current: StepLink | null;
+  skipTarget: StepLink | null;
+};
+
+export function getRelevantGuidedSteps(
+  data: InputData,
+  fromStep?: StepLink["id"]
+): OutputData {
+  const current = getNextGuidedStep(data, fromStep);
+
+  // There should be a relevant next step for every user (even if it's just
+  // going back to the dashbord), so we can't hit this line in tests (and
+  // shouldn't be able to in production either):
+  /* c8 ignore next 6 */
+  if (current === null) {
+    return {
+      current: null,
+      skipTarget: null,
+    };
+  }
+
+  return {
+    current: current,
+    skipTarget: getNextGuidedStep(data, current.id),
+  };
+}
+
+function getNextGuidedStep(
+  data: InputData,
+  fromStep?: StepLink["id"]
```

Comment:
Works great! The only thing I can think of right now is that it might be helpful to allow passing `Step["href"]` to `fromStep` alternatively as well. But I guess we can also just match and find the step id from the URL just before passing it in.

Reply:
Yeah, the reason I added the IDs in the first place is that links are so fiddly (do the include a trailing slash? are they relative? do they include the origin?). And since most calling code will know which step they're currently in, I figured it'd be safer to have them use the ID.